aim/sdk/sequence.py

```diff
@@ -14,6 +17,181 @@
+
+    def range(self, start, stop) -> 'SequenceData':
+        return SequenceV1Data(self.series_tree, columns=self.columns, step_range=(start, stop), n_items=self.n_items)
+
+    def sample(self, k) -> 'SequenceData':
+        return SequenceV1Data(self.series_tree, columns=self.columns, step_range=self.step_range, n_items=k)
+
+    def items_list(self) -> Tuple[List[int], List[Any]]:
+        iters = self._get_iters()
+        columns = [[] for _ in iters]
+        steps = []
+
+        if self.step_range is not None:
+            start, stop = self.step_range
+            if stop <= start or start < 0 or stop < 0:
+                return
+            for idx, val in iters[0]:
+                if idx < start:
+                    for it in iters[1:]:
+                        next(it)
+                    continue
+                if idx >= stop:
+                    break
+                steps.append(idx)
+                columns[0].append(val)
+                for it_index, it in enumerate(iters[1:]):
+                    columns[it_index + 1].append(next(it))
+        else:
+            for idx, val in iters[0]:
+                steps.append(idx)
+                columns[0].append(val)
+                for it_index, it in enumerate(iters[1:]):
+                    columns[it_index + 1].append(next(it))
+        length = len(steps)
+        step = (length // self.n_items or 1) if self.n_items > 0 else 1
+        if step != 1:
+            slice_ = slice(0, length, step)
+            last_step_needed = (length - 1) % step != 0
+            steps = steps[slice_] + [steps[-1]] if last_step_needed else steps[slice_]
+            for i, v in enumerate(columns):
+                columns[i] = v[slice_] + [v[-1]] if last_step_needed else v[slice_]
+
+        return steps, columns
+
+    def _get_iters(self) -> List[Iterator[Any]]:
+        iters = [self.arrays[0].items()]
+        for arr in self.arrays[1:]:
+            iters.append(arr.values())
+        return iters
+
+
+class SequenceV2Data(SequenceData):
+    def __init__(
+            self,
+            meta_tree, series_tree, *,
+            columns: List[Tuple[str, str]],
+            n_items: int = -1
+    ):
+        super().__init__(series_tree, version=2, columns=columns)
+        self.meta_tree = meta_tree
+        self.n_items = n_items
+        self.steps: ArrayView = self._get_array('step')
+
+    def view(self, columns: List[str]) -> 'SequenceData':
+        return SequenceV2Data(
+            self.meta_tree, self.series_tree, columns=self._checked_columns(columns), n_items=self.n_items)
+
+    def range(self, start, stop) -> 'SequenceData':
+        raise ValueError('Range selection cannot be applied to data stored with reservoir sampling.')
+
+    def sample(self, k) -> 'SequenceData':
+        return SequenceV2Data(self.meta_tree, self.series_tree, columns=self.columns, n_items=k)
+
+    def items_list(self) -> Tuple[List[int], List[Any]]:
```

Comment:
move implementation to `.numpy()` method to avoid multiple convertions from numpy to list and vice versa